src/schema/sources.ts

```diff
@@ -347,56 +374,146 @@ const sourceToGQL = (source: Source): GQLSource => ({
 });
 
 export enum SourcePermissions {
-  View,
-  Post,
-  Leave,
-  Delete,
-  Edit,
+  View = 'view',
+  Post = 'post',
+  PostLimit = 'post_limit',
+  PostDelete = 'post_delete',
+  MemberRemove = 'member_remove',
+  ModeratorAdd = 'moderator_add',
+  ModeratorRemove = 'moderator_remove',
+  InviteDisable = 'invite_disable',
+  Leave = 'leave',
+  Delete = 'delete',
+  Edit = 'edit',
 }
 
+export const roleSourcePermissions: Record<
+  SourceMemberRoles,
+  SourcePermissions[]
+> = {
+  owner: [
+    SourcePermissions.View,
+    SourcePermissions.Post,
+    SourcePermissions.PostLimit,
+    SourcePermissions.PostDelete,
+    SourcePermissions.MemberRemove,
+    SourcePermissions.ModeratorAdd,
+    SourcePermissions.ModeratorRemove,
+    SourcePermissions.InviteDisable,
+    SourcePermissions.Edit,
+    SourcePermissions.Delete,
+  ],
+  moderator: [
+    SourcePermissions.View,
+    SourcePermissions.Post,
+    SourcePermissions.PostDelete,
+    SourcePermissions.MemberRemove,
+    SourcePermissions.Edit,
+    SourcePermissions.Leave,
+  ],
+  member: [
+    SourcePermissions.View,
+    SourcePermissions.Post,
+    SourcePermissions.Leave,
+  ],
+};
+
+const requireGreaterAccessPrivilege: Partial<
+  Record<SourcePermissions, boolean>
+> = {
+  [SourcePermissions.MemberRemove]: true,
+};
+
+type BaseSourceMember = Pick<SourceMember, 'role'>;
+
+export const hasGreaterAccessCheck = (
+  loggedUser: BaseSourceMember,
+  member: BaseSourceMember,
+) => {
+  const memberRank = roleRank[member.role];
+  const loggedUserRank = roleRank[loggedUser.role];
+  const hasGreaterAccess = loggedUserRank > memberRank;
+
+  if (!hasGreaterAccess) {
+    throw new ForbiddenError('Access denied!');
+  }
+};
+
+interface SourcePermissionsProps {
+  permission: SourcePermissions;
+  validateRankAgainstId: string;
+}
+
+const hasPermissionCheck = (
+  source: Source | GQLSource,
+  member: BaseSourceMember,
+  permission: SourcePermissions,
+  validateRankAgainst?: BaseSourceMember,
+) => {
+  if (
```

Comment:
Can you explain this condition please?

Reply:
This was to support the previous implementation. This is to validate making a post and leaving a source that only happens on a Squad.

https://github.com/dailydotdev/daily-api/pull/1188/files/b94ea967f8264516644d8bbc8a267c36c2b5ed65#diff-95123b05e21b34d3705f1b3748fe295fa17db5caeeb22238a294368729175a07L378-L385

![image](https://user-images.githubusercontent.com/13744167/222960475-f4bef928-4913-4715-91fe-7286fb5bba02.png)

Reply:
I see, so let's be more explicit about this and change it to if it's the squad leader they can't leave. Very simple condition

Reply:
This was also to support avoiding making a post on a non-squad source. So I merged them. But I can definitely separate the two.

Reply:
Now with the arrays, these edge cases don't exist. You need to be a member to post, otherwise you will not have the permission. But please add or validate the relevant tests to make sure they don't exist indeed

Reply:
Got it. Will check the test cases if there are any existing checks and create one if there are none.